File: Kauailabs/demos/eclipse/DemoBot_Mecanum/Robot.cpp
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2017-2018 FIRST. All Rights Reserved.                        */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

#include <frc/TimedRobot.h>
#include <frc/Spark.h>
#include <frc/Encoder.h>
#include <frc/Joystick.h>
#include <frc/drive/MecanumDrive.h>
#include "AHRS.h"
using namespace frc;


class Robot : public TimedRobot {
public:
	Robot() {
		p_drive = new MecanumDrive(motor_left_front, motor_left_rear, motor_right_front, motor_right_rear);
		p_drive->SetRightSideInverted(true);
	}

	void AutonomousInit() override {
	}

	void AutonomousPeriodic() override {
	}

	void TeleopInit() override {}

	void TeleopPeriodic() override {
		double drive_strafe = -joy.GetX();
		double drive_fwd = -joy.GetY();
		double drive_rot = 0;
		if (joy.GetRawButton(6)) {
			drive_rot -= .8;
		}
		if (joy.GetRawButton(5)) {
			drive_rot += .8;
		}
		double gyro_angle = 0.0;
		p_drive->DriveCartesian(drive_strafe, drive_fwd, drive_rot, gyro_angle);
		/*
		wpi::outs() <<
				" Strafe:  " << drive_strafe <<
				" Fwd:  " << drive_fwd <<
				" Rot:  " << drive_rot <<
				" Gyro Angle:  " << gyro_angle <<
				" navX-Sensor Yaw:  " << ahrs.GetYaw() << "\n";
		*/
		/*
		motor_left_front.Set(drive_speed);
		motor_right_front.Set(drive_speed);
		motor_left_rear.Set(drive_speed);
		motor_right_rear.Set(drive_speed);
		int left_front_ticks = encoder_left_front.Get();
		int right_front_ticks = encoder_right_front.Get();
		int right_rear_ticks = encoder_right_rear.Get();
		int left_rear_ticks = encoder_left_rear.Get();
		wpi::outs() << "Joystick value:  " << drive_speed << " - encoders:  " <<
				left_front_ticks << " - " <<
				right_front_ticks << " - " <<
				right_rear_ticks << " - " <<
				left_rear_ticks << "\n";
		*/

	}

	void TestPeriodic() override {}

private:
	Joystick joy{0};
	Spark motor_left_front{0};
	Spark motor_right_front{1};
	Spark motor_right_rear{2};
	Spark motor_left_rear{3};
	Encoder encoder_left_front{0,1};
	Encoder encoder_right_front{2,3};
	Encoder encoder_right_rear{4,5};
	Encoder encoder_left_rear{6,7};
	MecanumDrive *p_drive;
	AHRS ahrs{50};
};

int main() {
	return frc::StartRobot<Robot>();
}

File: Kauailabs/demos/vscode/Demobot_Mecanum_AHRS/src/com/kauailabs/vmx/wpilib/Examples/DemoBot_Mecanum_AHRS/Robot.java
package com.kauailabs.vmx.wpilib.Examples.DemoBot_Mecanum_AHRS;

import edu.wpi.first.wpilibj.TimedRobot;
import edu.wpi.first.wpilibj.drive.MecanumDrive;
import edu.wpi.first.wpilibj.Spark;
import edu.wpi.first.wpilibj.Encoder;
import edu.wpi.first.wpilibj.Joystick;
import edu.wpi.first.wpilibj.SPI;
import com.kauailabs.navx.frc.AHRS;

public class Robot extends TimedRobot {

	Joystick joy;
	Spark motor_left_front;
	Spark motor_right_front;
	Spark motor_right_rear;
	Spark motor_left_rear;
	Encoder encoder_left_front;
	Encoder encoder_right_front;
	Encoder encoder_right_rear;
	Encoder encoder_left_rear;
	AHRS ahrs;
	MecanumDrive drive;
 
	public Robot() {
		joy = new Joystick(0);
		motor_left_front = new Spark(0);
		motor_right_front = new Spark(1);
		motor_right_rear = new Spark(2);
		motor_left_rear = new Spark(3);
		encoder_left_front = new Encoder(0,1);
		encoder_right_front = new Encoder(2,3);
		encoder_right_rear = new Encoder(4,5);
		encoder_left_rear = new Encoder(6,7);
		ahrs = new AHRS(SPI.Port.kMXP,(byte)50);
		drive = new MecanumDrive(motor_left_front, motor_left_rear, motor_right_front, motor_right_rear);
		drive.setRightSideInverted(true);

	}

	public void robotInit() {
		System.out.println("WPI Library Java DemoBot (Mecanum) Program");
	}

	public void teleopPeriodic() {
		System.out.println("teleopPeriodic");
		double drive_strafe = -joy.getX();
		double drive_fwd = -joy.getY();
		double drive_rot = 0;
		if (joy.getRawButton(6)) {
			drive_rot -= .8;
		}
		if (joy.getRawButton(5)) {
			drive_rot += .8;
		}
		
		float gyro_angle = ahrs.getYaw();
		drive.driveCartesian(drive_strafe, drive_fwd, drive_rot, gyro_angle);					
		System.out.println("Strafe:  " + drive_strafe + " Fwd:  " + 
			drive_fwd + " Rot:  " +
			drive_rot + " Gyro Angle:  " +
			gyro_angle + " navX-Sensor Yaw:  " +
			ahrs.getYaw());		
		
	}

	public void disabled() {
	}
}
